src/CommunityToolkit.Maui.Core/Views/Popup/MCTPopup.macios.cs

```diff
@@ -0,0 +1,176 @@
+	/// </summary>
+	public PageHandler? Control { get; private set; }
+
+	/// <summary>
+	/// An instace of the <see cref="IPopup"/>.
+	/// </summary>
+	public IPopup? VirtualView { get; private set; }
+
+	internal UIViewController? ViewController { get; private set; }
+
+	/// <summary>
+	/// Constructor of <see cref="MCTPopup"/>.
+	/// </summary>
+	/// <param name="mauiContext">An instace of <see cref="IMauiContext"/>.</param>
+	/// <exception cref="ArgumentNullException">If <paramref name="mauiContext"/> is null an exception will be thrown. </exception>
+	public MCTPopup(IMauiContext mauiContext)
+	{
+		this.mauiContext = mauiContext ?? throw new ArgumentNullException(nameof(mauiContext));
+	}
+
+	/// <summary>
+	/// Method to update the Popup's size.
+	/// </summary>
+	/// <param name="size"></param>
+	public void SetElementSize(Size size) =>
+		Control?.ContainerView?.SizeThatFits(size);
+
+	/// <inheritdoc/>
+	public override void ViewDidLayoutSubviews()
+	{
+		base.ViewDidLayoutSubviews();
+
+		_ = View ?? throw new InvalidOperationException($"{nameof(View)} cannot be null");
+		SetElementSize(new Size(View.Bounds.Width, View.Bounds.Height));
+	}
+
+	/// <summary>
+	/// Method to initialize the native implementation.
+	/// </summary>
+	/// <param name="element">An instance of <see cref="IPopup"/>.</param>
+	public void SetElement(IPopup element)
+	{
+		VirtualView = element;
+		ModalPresentationStyle = UIModalPresentationStyle.Popover;
+		CreateControl();
+		SetViewController();
+		SetPresentationController();
+		SetView();
+		AddToCurrentPageViewController();
+	}
+
+	void CreateControl()
+	{
+		_ = VirtualView ?? throw new NullReferenceException($"{nameof(VirtualView)} cannot be null.");
+
+		var view = (View?)VirtualView.Content;
+		var contentPage = new ContentPage { Content = view, BackgroundColor = Colors.Orange };
+
+		contentPage.Parent = Application.Current?.MainPage;
+		contentPage.SetBinding(VisualElement.BindingContextProperty, new Binding { Source = VirtualView, Path = VisualElement.BindingContextProperty.PropertyName });
+		Control = (PageHandler)contentPage.ToHandler(mauiContext);
+	}
+
+	void SetViewController()
+	{
+		Page currentPageRenderer;
+		var modalStackCount = Application.Current?.MainPage?.Navigation?.ModalStack?.Count ?? 0;
+		var stackCount = Application.Current?.MainPage?.Navigation?.NavigationStack?.Count ?? 0;
+		var mainPage = Application.Current?.MainPage ?? throw new NullReferenceException(nameof(Application.Current.MainPage));
+		if (modalStackCount > 0)
+		{
+			var index = modalStackCount - 1;
+			currentPageRenderer = mainPage.Navigation.ModalStack[index];
+		}
+		else if (stackCount > 0)
+		{
+			var index = stackCount - 1;
+			currentPageRenderer = mainPage.Navigation.NavigationStack[index];
+		}
+		else
+		{
+			currentPageRenderer = mainPage;
+		}
+
+		var viewController = (currentPageRenderer.Handler as PageHandler)?.ViewController;
```

Comment:
I think we should set controller only when it is null: https://github.com/xamarin/XamarinCommunityToolkit/blob/main/src/CommunityToolkit/Xamarin.CommunityToolkit/Views/Popup/iOS/PopupRenderer.ios.cs#L156

We had issue in XCT: https://github.com/xamarin/XamarinCommunityToolkit/pull/1762

Reply:
Good catch! Probably that was on `develop` branch (before v2) and I took the `main` as a reference. I'll update it ASAP